### PR TITLE
feat(developer): errs on unparseable regex 🙀

### PR DESCRIPTION
**Pull request by srl295**
- also, make sure NFD-incompliant ranges get mapped to escaped chars
- also, propagate UnicodeSet errs out

For: #10389

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/kmc-ldml/src/compiler/tran.ts [hunk 1]
```diff
@@ -141,13 +144,21 @@ export abstract class TransformCompiler<T extends TransformCompilerType, TranBas
...
 
     if (!sections?.meta?.normalizationDisabled) {
       // nfd here.
       cookedFrom = MarkerParser.nfd_markers(cookedFrom, true);
     }
 
+    try {
+      new RegExp(cookedFrom, 'ug');
+    } catch (e) {
+      this.callbacks.reportMessage(CompilerMessages.Error_UnparseableTransformFrom({ from: transform.from }));
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/tran.ts:158]: `new RegExp` can return useful information about the error; do we want to extract that?

```suggestion
      this.callbacks.reportMessage(CompilerMessages.Error_UnparseableTransformFrom({ from: transform.from, e }));
```

image

The exception messages returned are somewhat inconsistent -- some print the original regex, others don't.

File: developer/src/kmc-ldml/src/compiler/tran.ts [hunk 2]
```diff
@@ -168,14 +179,19 @@ export abstract class TransformCompiler<T extends TransformCompilerType, TranBas
       transforms: [],
       reorders: reorders.map(reorder => this.compileReorder(sections, reorder)),
     }
+    if (result.reorders.includes(null)) return null;
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/tran.ts:182]: Can you include a comment here?

File: developer/src/kmc-ldml/src/compiler/tran.ts [hunk 3]
```diff
@@ -141,13 +144,21 @@ export abstract class TransformCompiler<T extends TransformCompilerType, TranBas
...
-    this.checkRanges(cookedFrom); // check before normalizing
+    // cook the ranges also
+    cookedFrom = this.checkRanges(cookedFrom); // check before normalizing
 
     if (!sections?.meta?.normalizationDisabled) {
       // nfd here.
       cookedFrom = MarkerParser.nfd_markers(cookedFrom, true);
     }
 
+    try {
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/tran.ts:155]:
```suggestion
    // Verify that the regex is syntactically valid
    try {
```

File: developer/src/kmc-ldml/src/compiler/tran.ts [hunk 4]
```diff
@@ -200,12 +216,14 @@ export abstract class TransformCompiler<T extends TransformCompilerType, TranBas
     return defaults;
   }
 
-  private checkRanges(cookedFrom: string) {
+  /** true if ok */
```
Comment (mcdurdin) [inline developer/src/kmc-ldml/src/compiler/tran.ts:219]: I don't understand this comment

**Review by mcdurdin — APPROVED** [review]
LGTM, minor nits only

File: developer/src/kmc-ldml/src/compiler/tran.ts (hunk 1 above)
Comment (srl295) [reply developer/src/kmc-ldml/src/compiler/tran.ts:158]: that's a good idea. Might be challenging on the unit test side, though, if the wording changes or is localized.

Comment (mcdurdin) [reply developer/src/kmc-ldml/src/compiler/tran.ts]: unit tests probably shouldn't be checking the text of the error message, just the error code.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.267-alpha](https://keyman.com/downloads/releases/alpha/17.0.267)

**Pull request merged**